drivers: crypto: Add STM32 CRYP crypto driver

mrfuchs opened the pull request:
This patch adds crypto driver support for all STM32F4 devices providing
a CRYP peripheral.
This driver implements the AES ECB, CBC and CTR modes of operation.

It has been tested on a STM32F437 SoC running the Zephyr [crypto driver
sample](/zephyrproject-rtos/zephyr/tree/master/samples/drivers/crypto):

```
> [00:00:05.645,000] <inf> main: Cipher Sample
> [00:00:05.915,000] <inf> main: ECB Mode
> [00:00:06.161,000] <inf> main: Output length (encryption): 16
> [00:00:06.516,000] <inf> main: ECB mode ENCRYPT - Match
> [00:00:06.842,000] <inf> main: Output length (decryption): 16
> [00:00:07.197,000] <inf> main: ECB mode DECRYPT - Match
> [00:00:07.522,000] <inf> main: CBC Mode
> [00:00:07.768,000] <inf> main: Output length (encryption): 80
> [00:00:08.123,000] <inf> main: CBC mode ENCRYPT - Match
> [00:00:08.449,000] <inf> main: Output length (decryption): 64
> [00:00:08.804,000] <inf> main: CBC mode DECRYPT - Match
> [00:00:09.130,000] <inf> main: CTR Mode
> [00:00:09.375,000] <inf> main: Output length (encryption): 64
> [00:00:09.730,000] <inf> main: CTR mode ENCRYPT - Match
> [00:00:10.056,000] <inf> main: Output length (decryption): 64
> [00:00:10.411,000] <inf> main: CTR mode DECRYPT - Match
> [00:00:10.737,000] <inf> main: CCM Mode
> [00:00:10.982,000] <err> crypto_stm32: Unsupported mode
```

AES-CCM mode is omitted from the driver, as ST's CRYP peripheral seems
to require input data is passed as full AES blocks, i.e. its length is
a multiple of 128 bits. Otherwise encryption works, but the generated
authentication tag does not match, although the CRYP HAL module claims
to have a [workaround](/zephyrproject-rtos/hal_stm32/blob/566f4eb616f9e61bec01aca9918cace4b01e11b5/stm32cube/stm32f4xx/drivers/src/stm32f4xx_hal_cryp.c#L4391-L4397) for this case.

mrfuchs committed:
dts: arm: stm32f4: Add CRYP support for STM32F4xx SoC series

Add Cryptographic Accelerator support for STM32F415/417 and STM32F43x
devices.

Signed-off-by: Markus Fuchs <markus.fuchs@de.sauter-bc.com>

mrfuchs committed:
drivers: crypto: Add STM32 CRYP crypto driver

This patch adds crypto driver support for all STM32F4 devices providing
a CRYP peripheral.
This driver implements the AES ECB, CBC and CTR modes of operation.

It has been tested on a STM32F437 SoC running the Zephyr crypto driver
sample.

Signed-off-by: Markus Fuchs <markus.fuchs@de.sauter-bc.com>

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -0,0 +1,400 @@
+	if ((mode != CRYPTO_CIPHER_MODE_ECB) &&
+	    (mode != CRYPTO_CIPHER_MODE_CBC) &&
+	    (mode != CRYPTO_CIPHER_MODE_CTR)) {
+		LOG_ERR("Unsupported mode");
```
Please add a status clarifying CCM support.

mrfuchs replied:
Thanks for your feedback.

Do you mean adding a comment that CCM is not supported, like

> Due to hardware restrictions CCM mode is not supported.

or a comment explaining why CCM is not supported, like

> The CRYP peripheral supports the AES ECB, CBC, CTR, CCM and GCM modes of operation, of which ECB, CBC, CTR and CCM are supported through the crypto API. However, in CCM mode, although the STM32Cube HAL driver follows the documentation (cf. RM0090, par. 23.3) by padding incomplete input data blocks in software prior encryption, incorrect authentication tags are returned for input data which is not a multiple of 128 bits. Therefore, CCM mode is not supported by this driver.

erwango replied:
Wait, @tbistm may have a hint to overcome the restriction.

tbistm replied:
CCM can malfunction for several reasons. A wrong Tag may be due to :
1) a Tag with a variable length (not 16 bytes). It can be in theory (from NIST, A.1 Length Requirements) an element of {4, 6, 8, 10, 12, 14, 16} (bytes). but CRYP driven systematically reports its maximum size because it supposes memory is always reserved by upper layers.  
2) data alignment (including B0 and B1 should be 4 bytes align)

mrfuchs replied:
So, the CRYP peripheral does not natively support the test vector used in the crypto sample, does it?
https://github.com/zephyrproject-rtos/zephyr/blob/1ab39e7c52237d5c438e11b59e2f58e00cba222e/samples/drivers/crypto/src/main.c#L366-L381
https://github.com/zephyrproject-rtos/zephyr/blob/1ab39e7c52237d5c438e11b59e2f58e00cba222e/samples/drivers/crypto/src/main.c#L392-L397
Is there any way to support partial inputs and tags with lengths other than 16 bytes?

There are other STM32 crypt drivers, I am not sure if they support other configurations:
- The [Linux driver](https://github.com/torvalds/linux/blob/0c99ee44b8921b872089a7374c733310d995fd92/drivers/crypto/stm32/stm32-cryp.c#L1158-L1175) at least looks like it can handle authentication tags smaller 16 bytes, but it targets the [STM32F756 and STM32MP1 SoCs](https://github.com/torvalds/linux/blob/0c99ee44b8921b872089a7374c733310d995fd92/drivers/crypto/stm32/stm32-cryp.c#L1907-L1909) only.
- The [wolfSSL driver](https://github.com/wolfSSL/wolfssl/blob/6b4551c012941ebc407d843c9a6dc835da546a34/wolfcrypt/src/aes.c#L5620-L5626) claims to generate the GCM "Auth correctly for partial or when IV != 12". Of course, GCM authentication tag generation differs from the one for CCM.

mrfuchs replied:
I forgot the STM32/X-CUBE Crypto Library. It is [NIST CAVP-certified](https://www.st.com/content/ccc/resource/sales_and_marketing/presentation/product_presentation/group0/79/b4/63/20/21/53/47/54/fips_cavp_certification/files/fips_cavp_certification.pdf/jcr:content/translations/en.fips_cavp_certification.pdf), uses the CRYP peripheral for AES (cf. [table 4 on page 26 in UM1924](https://www.st.com/content/ccc/resource/technical/document/user_manual/group0/f9/6e/f2/a2/b4/ec/49/c0/DM00215061/files/DM00215061.pdf/jcr:content/translations/en.DM00215061.pdf#page=26)) on the mentioned STM32F4 devices, supports nonces with sizes from 7 to 13 bytes, authentication tags with sizes from 4 to 16 bytes, and should work on complete and partial input data.

tbistm replied:
I suggest to limit the SW without CCM. Your comment is relevant.
There might be a problem and we have been investigating.

mrfuchs replied:
> Please add a status clarifying CCM support.

Comment added.

erwango replied:
Sorry fro late feedback, can you pout your later proposition instead? (the detailed one)
I'd like to avoid that someone reads this and thinks this was a mistake after a quick check in the ref manual.

mrfuchs replied:
Comment has been changed to the detailed one.

mrfuchs replied:
> There might be a problem and we have been investigating.

Thanks. Please keep me posted.

erwango reviewed (changes requested):
One minor comment.
Then, it would be good to enable driver on a in tree board so test could be run or at least built.
For now, only upstream board supporting the IP is mikroe_mini_m4_for_stm32, can you add it to `samples/drivers/crypto/sample.yaml` with `build_only: true`. Can you add that ?

erwango commented:
cc @TBISTM

tbistm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -0,0 +1,400 @@
+	}
+
+	if ((ctx->keylen != 16U) &&
+	    (ctx->keylen != 24U) &&
```
Note that a 192 bits key is optional and may not be supported by all HW crypto IP.
Unfortunately, there is currently no HAL API to know this capability on the runtime. It should be statically managed.
This implementation should be fine for F4 but this should be updated to have a code more generic addressing all the ST families.

mrfuchs replied:
Thanks for pointing this out. I'll add a comment to this line, so this will later be configurable via a device tree option.

tbistm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -0,0 +1,400 @@
+
+static struct crypto_stm32_data crypto_stm32_dev_data = {
+	.hcryp = {
+		.Instance = CRYP
```
It is a simple remark (relevant only for a code generic on all the families)
CRYP is the common name of the instance and this is fine for F4.
Nevertheless, this wording is different on other families (e-g L5 series). It is called : AES
I do not know why ?!

This is conditionally driven by define in : stm32hxxxx.h :
#define CRYP                ((CRYP_TypeDef *) CRYP_BASE)

mrfuchs replied:
Yes. This was done on purpose. I decided against `.Instance = (CRYP_TypeDef *) DT_CRYP_0_BASE_ADDRESS`, because I wanted to make clear, that - at this point - only the Cryptographic Accelerator peripheral is supported, the AES hardware encryption accelerator is not. That could easily be added at a later time, I guess. As I don't have suitable hardware at the moment, I decided to concentrate on CRYPT.

tbistm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -0,0 +1,400 @@
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	session->in_use = false;
+
```
After free I would expect  HW to be kept under reset, or its clock to be shut down ? 
Here the IP will still be running ...

mrfuchs replied:
Should we loop over all sessions here and call `HAL_CRYP_DeInit`, if no more session is in use?

tbistm replied:
I think so. 
HAL_CRYP_DeInit() will call HAL_CRYP_MspDeInit(hcryp). 
This will shutdown the IP.
Check that the following is called at the end :   
  __HAL_RCC_CRYP_FORCE_RESET();
  __HAL_RCC_CRYP_RELEASE_RESET();

mrfuchs replied:
I followed your advice in the latest commit. Thanks.

tbistm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -0,0 +1,400 @@
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	status = HAL_CRYP_SetConfig(&data->hcryp, &session->config);
```
I would have expected HAL_CRYP_SetConfig() outside do_encrypt() because of setting once should be enough.

mrfuchs replied:
I wanted to avoid code duplication, so I moved `HAL_CRYP_SetConfig` from the `crypto_stm32_XXX_XXcrypt` functions to `do_XXcrypt`. Moving it outside the crypt functions wouldn't work well in a multi-session scenario, I guess.

tbistm replied:
Ok.

mrfuchs committed:
samples: drivers: crypto: Add STM32 driver support

Add STM32 CRYP driver support and a corresponding build-only test to
the crypto sample project.

Signed-off-by: Markus Fuchs <markus.fuchs@de.sauter-bc.com>

mrfuchs commented:
> For now, only upstream board supporting the IP is mikroe_mini_m4_for_stm32, can you add it to `samples/drivers/crypto/sample.yaml` with `build_only: true`. Can you add that ?

This has been added now. Can you double-check, please?

erwango commented on drivers/crypto/Kconfig.stm32:
```diff
@@ -0,0 +1,24 @@
+config CRYPTO_STM32_MAX_SESSION
+	int "Maximum of sessions STM32 crypto driver can handle"
+	default 2
+	depends on CRYPTO_STM32
```
This line has no effect since redundant with "if CRYPTO_STM32"

mrfuchs replied:
This line has been removed.

erwango commented on samples/drivers/crypto/prj_stm32.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_MAIN_STACK_SIZE=4096
+
+CONFIG_CRYPTO=y
+CONFIG_CRYPTO_STM32=y
```
This line should not be required as CONFIG_CRYPTO_STM32 should be enabled when CONFIG_CRYPTO is enabled

mrfuchs replied:
This line has been removed.

erwango reviewed (changes requested):
Few additional comments

galak commented on samples/drivers/crypto/sample.yaml:
```diff
@@ -27,3 +27,9 @@ tests:
+  sample.drivers.crypto.stm32:
+    tags: crypto
+    platform_whitelist: mikroe_mini_m4_for_stm32
+    filter: dt_compat_enabled("st,stm32-cryp")
```
Why do we need both `filter` and `platform_whitelist`?

mrfuchs replied:
Removed filter.

erwango replied:
@galak, @ulfalizer genuine question: how is `dt_compat_enabled` supposed to work as sanitycheck filter?
When running sanitycheck, do we build all boards.dts to see if "compat,foo' is actually enabled ? (This is not what I'm seeing at least)
Or is the filter applied after platforms builds and then if filter does not match, the configuration is skipped? (which would explain the mysterious "skipped").

In any case, it seems that this filter is missing something as `sample.drivers.crypto.stm32` is applied on bunch of other platforms ....

galak commented on soc/arm/st_stm32/stm32f4/dts_fixup.h:
```diff
@@ -523,4 +523,12 @@
 #define DT_CAN_1_CLOCK_BUS		DT_ST_STM32_CAN_40006400_CLOCK_BUS
 #define DT_CAN_1_CLOCK_BITS		DT_ST_STM32_CAN_40006400_CLOCK_BITS
 
+#define DT_CRYP_0_NAME			DT_INST_0_ST_STM32_CRYP_LABEL
```
Can the driver not use `DT_INST_` defines directly and not require any dts_fixup.h changes?

mrfuchs replied:
Removed dts_fixup.h changes.

galak reviewed (changes requested):
Some minor DT/test questions/changes.

erwango commented:
@mrfuchs can you check @galak 's comments so we cant get this in for V2.2?

erwango commented on samples/drivers/crypto/sample.yaml:
```diff
@@ -27,3 +27,8 @@ tests:
         - ".*: CCM Mode"
+  sample.drivers.crypto.stm32:
+    tags: crypto
+    platform_whitelist: mikroe_mini_m4_for_stm32
```
I would have preferred filter, so we can easily add more targets. 
Any reason why you preferred whitelist ?

mrfuchs replied:
The filter used before didn't work (project was built on multiple targets and failed for all but the mikroe_mini_m4_for_stm32). Can you share the filter expression you have in mind?

erwango replied:
Actually, I would have expected the filter to work.
Let's stick with the white list for now.
